bacpop/beebop_py
# bacpop-100 External clusters

This branch ensures that clusters defined in the GPSC external clusters file are returned for all analysis results, rather than internal (poppunk) clusters. 

This branch is deployed to http://beebop-dev.dide.ic.ac.uk
Nick has had a look at this and declared it to be working as expected. 

Key changes:
- added the external clusters csv file to the resources folder
- pass that filepath to poppunk assign and `generate_visualisation` calls
- use the external_clusters csv file (rather than clusters.csv) generated by the assign call to build the API response
- also use the external_clusters file as `previous_query_clustering` parameter of the poppunk `generate_visualisation` calls
- when constructing assign responses, we want to assign a single external cluster to each sample, whereas the external clusters files have multiple cluster numbers, separated by semicolons. As requested by Nick, we assign lowest cluster number for each sample. 
- we also put "GPSC" prefix on the clusters, to be displayed in the front end
- there is one place where we still need to deal with internal poppunk clusters, and that's when passing the include file to the `generate_visualisation` call for microreact. That include file is generated by the `summarise_clusters` call, and is tagged with the internal cluster. So we save a mapping of external cluster labels to internal cluster number when we do the assignment, for later use to fetch this file on create microreact. 
- there was as issue when mapping clusters to network components (in `utils.generate_mapping`) - network components get arbitrary ids which we need to map to clusters ourselves. Previously we were abe to assume that a cluster which shared any sample with a component could be mapped to that component. However, the network now seems to be generating some non-cluster single-sample components which we definitely don't want to map to. However, we can't require that components and clusters be exact matches either (possibly because of the lowest number assignment?) so instead I've put in some very inefficient best-match code which seems to work well enough. Nick thinks this is an issue with the poppunk code, and is going to look into it, so we should be able to revert this later. 
- I've tried to rationalise the tests a bit so that they share code to do assignment and run network, and so tests don't rely on each others' output.

No changes were required in beebop. The GPSC prefix makes the network tab labels wrap, which looks a bit weird, but I think we can sort that out in the new UI!

To test:
You should find that the attached files (provided by Nick): 
[19183_3#105.fasta.txt](https://github.com/bacpop/beebop_py/files/14613272/19183_3.105.fasta.txt) gets assigned to GPSC22
[22841_2#137.fasta.txt](https://github.com/bacpop/beebop_py/files/14613274/22841_2.137.fasta.txt) gets assigned to GPSC9

..and that network and download output look right, and are tagged with the external cluster number. 
(NB remove .txt suffix to upload to beebop - github won't attach a .fasta file!)
